#!/bin/env bash

################################################################################################################
### HELP -------------------------------------------------------------------------------------------------------
################################################################################################################
script_name='ScanMotif.sh'

# Get user id for custom manual pathways
usr=`id | sed -e 's@).*@@g' | sed -e 's@.*(@@g'`

# Text font variables
END='\033[0m'
BOLD='\033[1m'
UDL='\033[4m'

# Show command manual
Help()
{
echo -e "${BOLD}####### INDEXBAM MANUAL #######${END}\n\n\
${BOLD}SYNTHAX${END}\n\
	sh ${script_name} [options] <fasta_file> <motif_file>\n\n\

${BOLD}DESCRIPTION${END}\n\
	Look for a specific DNA motif on the entire genome.

${BOLD}OPTIONS${END}\n\
	${BOLD}-F${END} ${UDL}string${END}, Output${BOLD}F${END}ormat\n\
		Define format of output with corresponding columns.\n\
		If 'bed' is selected, output file will have 6 columns : chr, start, end, motif name, log-odds score (floored to an integer), strand\n\
		If 'txt' is selected, output file will have 7 columns : Site ID (motif name + number), chr, start, end, strand, log-odds score (floored to an integer), sequence\n\		
		Default = 'txt'\n\n\
	
${BOLD}ARGUMENTS${END}\n\
	${BOLD}<fasta_file>${END}\n\
		Path to genome reference FASTA file.\n
		It can usually be downloaded from Ensembl genome browser.\n\n\
	${BOLD}<motif_file>${END}\n\
		Path to .motif file containing DNA motif to look for in provided genome.\n\
		This type of file is generated by HOMER or available online.\n\n\
   		
${BOLD}EXAMPLE USAGE${END}\n\
	sh ${script_name} ${BOLD}-O${END} 'unused' ${BOLD}${usr}/Ref/Genome/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa ${usr}/Ref/Motifs/Factor.motif${END}\n"
}

################################################################################################################
### OPTIONS ----------------------------------------------------------------------------------------------------
################################################################################################################

# Set default values
F_arg='txt'

# Change default values if another one is precised
while getopts ":O:" option; do
	case $option in
		F) # OUTPUT FORMAT
			F_arg=${OPTARG};;
		\?) # Error
			echo "Error : invalid option"
			echo "      Allowed options are [-F]"
			echo "      Enter 'sh ${script_name} help' for more details"
			exit;;
	esac
done

# Checking if provided option values are correct
case $F_arg in
	TXT|Txt|txt) 
		F_arg=''
		file_ext='txt;;
	BED|Bed|bed) 
		F_arg='-bed '
		file_ext='bed';;
	*) 
		echo "Error value : -S argument must be 'true' or 'false'"
		exit;;
esac 

# Deal with options [-O] and arguments [$1|$2|...]
shift $((OPTIND-1))

################################################################################################################
### ERRORS -----------------------------------------------------------------------------------------------------
################################################################################################################

if [ $# -eq 1 ] && [ $1 == "help" ]; then
	Help
	exit
elif [ $# -lt 2 ]; then
	# Error if arguments are missing
	echo "Error synthax : please use following synthax"
	echo "      sh ${script_name} [options] <fasta_file> <motif_file>"
	exit
fi

################################################################################################################
### SCRIPT -----------------------------------------------------------------------------------------------------
################################################################################################################

## SETUP - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
module load homer/4.11

# Generate REPORT
echo '#' >> ./0K_REPORT.txt
date >> ./0K_REPORT.txt

Launch()
{
# Launch COMMAND and save report
echo -e "#$ -V \n#$ -cwd \n#$ -S /bin/bash \n"${COMMAND} | qsub -N ${JOBNAME} ${WAIT}
echo -e ${JOBNAME} >> ./0K_REPORT.txt
echo -e ${COMMAND} | sed 's@^@   \| @' >> ./0K_REPORT.txt
}
WAIT=''

## SCAN MOTIF - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

# Set variables for jobname
fasta=`echo ${1} | sed -e 's@.*\/@@g' | sed -e 's@\.fa\|\.fasta@@g'`
motif=`echo ${2} | sed -e 's@.*\/@@g' | sed -e 's@\.motif@@g'`

# Define JOBNAME and COMMAND considering WAIT
JOBNAME="ScanMotif_${motif}_in_${fasta}"
COMMAND="scanMotifGenomeWide.pl ${2} ${1} ${F_arg}> ${motif}_in_${fasta}.${file_ext}"
Launch

